### ISSUE TITLE
My one day review

### ISSUE THREAD

**Issue by jbogaardt** [casact/tryangle#1]
I've had a lot of fun exploring this package and its very well done.  Its such a familar environment for an actuary to learn more about the ML/DL toolkit.  You're doing fantastic work.  Here are some critiques (that shouldn't be taken as negative in the slightest):

### The API
The one thing that catches me off guard is that `sample_weight` gets embedded in `X`. This is different to how `sklearn` and `chainladder` approach it.  In most cases you keep very close to `sklearn` (right down to the nested imports) so I'm curious of the reason for this departure. This also seems to be the reason you have to maintain your own subclasses of `chainladder` estimators which can get tedious to keep synced.  Is `TryangleData` really a necessary abstraction?

### The (1,1,k,k) shape requirement
It fails for me on an asymmetric triangles. For example, using this `X` in place of those in the `AutoEnsemble` example, I run into error:
```python
tri = cl.load_sample('quarterly')['paid']
X = tryangle.TryangleData(tri, sample_weight=tri.latest_diagonal*0+2000)
```
I hope its not a fault of `chainladder`  😮.

It's not clear to me why more than one index cannot be supported.  Is this just because this is version `0.1.0` and I am just being too eager?  It seems the CDR minimization should just work for a higher dimensioned `Triangle`.  I do think relaxing this requirement in the future would help move the package more towards practical real world application.  Something I'm currently working on in chainladder==0.8.4 is `groupby` hyperparameters.  Refer to [this example](https://github.com/casact/chainladder-python/issues/153). It would be really slick to add this hyperparameter to the search space, but on a (1,1,k,k) triangle, is not so useful.

Another common usage pattern that deserves thought is blending paid and incurred methods together. Since these focus on different columns, this seemingly basic actuarial "ensembling" practice currently is not supported. 

I do want to read your paper again and validate the intuition behind the optimized parameters the library recommends for me and I'm sure I'll have more (hopefully) useful feedback for you once I do that.

Overall, its very nicely done, much more mature than the `0.1.0` version tag you have on it!

**Comment by cbalona** [comment]
Thank you very much for the feedback, John.

Here are my comments:

> The one thing that catches me off guard is that sample_weight gets embedded in X. This is different to how sklearn and chainladder approach it. In most cases you keep very close to sklearn (right down to the nested imports) so I'm curious of the reason for this departure. This also seems to be the reason you have to maintain your own subclasses of chainladder estimators which can get tedious to keep synced. Is TryangleData really a necessary abstraction?

I have not rethought this for a while, but I believe it still is necessary. The issue relates to how `Triangle`'s are indexed. `Triangle`'s are 4-dimenstional, and sklearn is designed to work with 1 dimenstional arrays. Deep within the `BaseSearchCV`  there is a private function `_fit_and_score` which fits and scores across folds using `_safe_indexing`. I created `TryangleData` to have the functionality to index using a list of integers which is compatible with `_safe_indexing`. So if we imagine a (1, 1, k, k) `Triangle`, index 0 is the top-left-most cell, and k * k is the bottom-right-most cell. I then index the `Triangle` using an array of integers to get sub-triangles.

Ways around this are to have `Triangle`'s themselves indexable by a list of integers, or making `_safe_indexing` compatible with `Triangle`'s. I think the first option is better, but I am not sure how this would impact `chainladder` overall, and I have not considered multi-dimensional triangles. Which leads me to the next point.

> It's not clear to me why more than one index cannot be supported. Is this just because this is version 0.1.0 and I am just being too eager? It seems the CDR minimization should just work for a higher dimensioned Triangle. I do think relaxing this requirement in the future would help move the package more towards practical real world application. Something I'm currently working on in chainladder==0.8.4 is groupby hyperparameters. Refer to this example. It would be really slick to add this hyperparameter to the search space, but on a (1,1,k,k) triangle, is not so useful.

I agree more than one index would be beneficial, and certainly is something I want to look at in future. One thing the prevented  me from implementing this immediately is that I was not sure how we'd do parameter selection then. Typically, if we fit a Chainladder as a simple example, we'd select one set of parameters for the triangle. Say, `drop_high = True` and `n_periods = 10`. If we have 10 LOB's in the triangle, would we not practically select parameters for each triangle separately instead of applying the same set for all? I guess we could say that each triangle gets optimised individually and then a list of optimal parameters, one for each LOB, is provided. But then that's just abstracting a loop into a handler function.

> Another common usage pattern that deserves thought is blending paid and incurred methods together. Since these focus on different columns, this seemingly basic actuarial "ensembling" practice currently is not supported.

This is great and I would like to explore this more. Perhaps this is the primary use case for more than one index?

> It fails for me on an asymmetric triangles. For example, using this X in place of those in the AutoEnsemble example, I run into error:

I will look into this more deeply when I can. I had not tried asymmetrical triangles yet, but, it certainly should work.

Thanks again for the feedback, John!

**Comment by jbogaardt** [comment]
> Ways around this are to have `Triangle`'s themselves indexable by a list of integers, or making `_safe_indexing` compatible with `Triangle`'s. I think the first option is better, but I am not sure how this would impact `chainladder` overall, and I have not considered multi-dimensional triangles. Which leads me to the next point.

I don't think adding this would affect any of the existing functionality of the chainladder.Triangle since it is currently unimplemented and unused functionality.  I'd be happy to support it if it reduces work-arounds on your end. Getting closer to `sklearn` compliance is a reasonable goal for the chainladder library regardless.

**Comment by cbalona** [comment]
> I don't think adding this would affect any of the existing functionality of the chainladder.Triangle since it is currently unimplemented and unused functionality. I'd be happy to support it if it reduces work-arounds on your end. Getting closer to `sklearn` compliance is a reasonable goal for the chainladder library regardless.

That’s great then! I do just need to double check if that will solve the problem. Let me fork and add the functionality to chainladder and run my tests. I think there is still an issue with how sklearn treats the sample_weight which might be the other reason I combined them, but can’t remember at the moment.

**Comment by jbogaardt** [comment]
> One thing the prevented me from implementing this immediately is that I was not sure how we'd do parameter selection then. Typically, if we fit a Chainladder as a simple example, we'd select one set of parameters for the triangle. Say, `drop_high = True` and `n_periods = 10`. If we have 10 LOB's in the triangle, would we not practically select parameters for each triangle separately instead of applying the same set for all?

I've thought about this a little more, and the more I think about it, this ambiguity is what the `groupby` parameter in `chainladder==0.8.4` addresses.  Consider this example:

```python
import chainladder as cl
clrd = cl.load_sample('clrd')['CumPaidLoss']
dev = cl.Development().fit(clrd)
```
By itself, it produces (with error due to malformed triangles), a set of independent patterns for each "row" of the triangle index.  A more sensible approach would be to share development patterns (say for each LOB).  In earlier versions of chainladder this would be done as follows:

```python
import chainladder as cl
clrd = cl.load_sample('clrd')['CumPaidLoss']
dev = cl.Development().fit(clrd.groupby('LOB').sum()).transform(clrd)
```
The problem with this approach is that fitting LDFs at the LOB grain, is not actually specified as a hyperparameter (or part of the model definition, which I think it should).  This makes such a workflow incompatible with  `Pipeline` or a `GridSearch` or `RandomizedSearch` since we are fitting and transforming two different triangles.

However, with chainladder==0.8.4, this can now be reworked with this `Pipeline` compatible approach:
```python
import chainladder as cl
clrd = cl.load_sample('clrd')['CumPaidLoss']
dev = cl.Development(groupby='LOB').fit_transform(clrd)
```
In this case, the ambiguity you highlighted is handled by the `groupby` parameter.  In other words, with 0.8.4, `Tryangle`, shouldn't necessarily have to worry about how to handle the index. It should be decided by the end users workflow as part of their model definition.

**Comment by jbogaardt** [comment]
Never mind, I think I misread your comment. In those cases where you want different hyperparameters for each triangle, you'd want an index of length 1 and run independent optimizations. 

Still, the use case I would suggest for the groupby feature would be having Tryangle pick from the best grouping of triangles. Maybe LOB is an appropriate grouping, maybe geography, or maybe "cause of loss" is. Assuming a triangle has these features in its index, then one could specify and test various groupby parameters in the optimization and Tryangle could tell you the optimal aggregation of granular data.

**Comment by cbalona** [comment]
Thanks John. Yes I think it would be useful to be able to optimize the groupby parameter too so it is great that it is now a development parameter. So your pipeline would begin with one multidimensional triangle, then split based on the grouping into multiple triangles that you'd then score, and through that select the best groupby parameter. Then a separate step will be to individually optimise each group of the initial triangle.
